playbooks/common/openshift-master/config.yml

```diff
@@ -337,12 +361,54 @@
     with_items: openshift_master_named_certificates
     when: named_certs_specified | bool
 
-- name: Configure master instances
-  hosts: oo_masters_to_config
+- name: Configure first master instance
```

Comment:
> What if I consider openshift_version an override instead of something that's almost always set, and push more of the logic into openshift_docker role.

As far as the roles are concerned, I believe this should always be the case.

> config playbook for first master passes no openshift_version
openshift_docker role inherits responsibility of sorting out the various inventory vars and setting openshift_version:
if openshift_version passed use it (still converts generic 3.2 to specific 3.2.0.44 if necessary)
else use installed version if present
else use inventory vars if present (release / pkg_version / image_tag)
only openshift_docker will reference these, everything else uses openshift_version
else use lookup latest and use that

What is the plan for looking up the latest with containerized installs? Just pull the `latest` tag? If so, I'm good with that approach, we'd just need to keep in mind that we'd need to take care when testing this internally where the latest tags are broken (which I'm fine with, we'd just need to make sure it's documented somewhere).

> config playbook for subsequent masters and nodes will pass openshift_version = first master installed version fact
upgrade playbook will need to determine always set openshift_version.
Now the role defaults to protecting installed version, and installing latest if nothing is requested. It involves no upgrade flag, but orchestrator can request a specific version that takes over everything for upgrades and/or pinning to master version.

> We do have to choose one of (a) protect installed versions and (b) sync everything to first master. Which one will customers prefer?

> I believe the above can solve both cases, we'd just have the config playbook also check for pre-existing version and pass it as openshift_version, otherwise pass first master.

I can live with that, though doing that in a way that wouldn't require pre-loading the host facts may prove difficult. The easiest way I can think of to do that would be to use `vars` on the role invocation, in which case the template would be evaluated at the time it is used by a task within the role.

The alternative could be that to get 'a', the user would need to explicitly set the version variable.